rclcpp/include/rclcpp/create_timer.hpp

```diff
@@ -132,12 +115,97 @@ create_wall_timer(
             "Casting timer period to nanoseconds resulted in integer overflow."};
   }
 
+  return period_ns;
+}
+
+/// Convenience method to create a wall timer with node resources.
+/**
+ *
+ * \tparam DurationRepT
+ * \tparam DurationT
+ * \tparam CallbackT
+ * \param period period to execute callback. This duration must be 0 <= period < nanoseconds::max()
+ * \param callback callback to execute via the timer period
+ * \param group callback group
+ * \param node_base node base interface
+ * \param node_timers node timer interface
+ * \return shared pointer to a wall timer
+ * \throws std::invalid_argument if either node_base or node_timers
+ * are null, or period is negative or too large
+ */
+template<typename DurationRepT, typename DurationT, typename CallbackT>
+typename rclcpp::WallTimer<CallbackT>::SharedPtr
+create_wall_timer(
+  std::chrono::duration<DurationRepT, DurationT> period,
+  CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group,
+  node_interfaces::NodeBaseInterface * node_base,
+  node_interfaces::NodeTimersInterface * node_timers)
+{
+  if (node_base == nullptr) {
+    throw std::invalid_argument{"input node_base cannot be null"};
+  }
+
+  if (node_timers == nullptr) {
+    throw std::invalid_argument{"input node_timers cannot be null"};
+  }
+
+  const std::chrono::nanoseconds period_ns = safe_cast_to_period_in_ns(period);
+
+  // Add a new wall timer.
   auto timer = rclcpp::WallTimer<CallbackT>::make_shared(
     period_ns, std::move(callback), node_base->get_context());
   node_timers->add_timer(timer, group);
   return timer;
 }
 
+/// Convenience method to create a general timer with node resources.
+/**
+ *
+ * \tparam DurationRepT
+ * \tparam DurationT
+ * \tparam CallbackT
+ * \param period period to execute callback. This duration must be 0 <= period < nanoseconds::max()
+ * \param callback callback to execute via the timer period
+ * \param group callback group
+ * \param node_base node base interface
+ * \param node_timers node timer interface
+ * \return shared pointer to a generic timer
+ * \throws std::invalid_argument if either node_base, node_timers or node_clock
+ * are null, or period is negative or too large
+ */
+template<typename DurationRepT, typename DurationT, typename CallbackT>
+typename rclcpp::GenericTimer<CallbackT>::SharedPtr
+create_timer(
+  std::chrono::duration<DurationRepT, DurationT> period,
+  CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group,
+  node_interfaces::NodeBaseInterface * node_base,
+  node_interfaces::NodeTimersInterface * node_timers,
+  node_interfaces::NodeClockInterface * node_clock)
+{
+  if (node_base == nullptr) {
+    throw std::invalid_argument{"input node_base cannot be null"};
+  }
+
+  if (node_timers == nullptr) {
+    throw std::invalid_argument{"input node_timers cannot be null"};
+  }
+
+  if (node_clock == nullptr) {
+    throw std::invalid_argument{"input node_clock cannot be null"};
+  }
```

Comment:
can you add this checks to the `create_timer` function in L38 and reuse it?
Thanks!

It would also be nice to reorder so you first have all `create_timer()` overloads and then you have `create_wall_timer()`

Reply:
This is unfortunately not trivial because it involves a function signature change. The reason is because there is a mismatch between argument types (some are raw pointers, and others are shared_ptr). What you have asked for requires a raw -> shared_ptr conversion, which is not possible.

Reply:
@ivanpauno  please take a look at commit https://github.com/ros2/rclcpp/pull/1975/commits/329ec5215c39b62cb395d4644751a25b5d84b86b. It's a fairly large change that I think achieves what you have asked for. Notably it:
- Removes raw pointers in favor of shared pointers.
- Forces the create_timer call to follow a sequence of function calls.
- Moves the argument nullptr checks to a common place.
It did require making changes to some other ROS2 components to avoid passing around a raw pointer from which a shared pointer cannot be derived.

Reply:
> This is unfortunately not trivial because it involves a function signature change. The reason is because there is a mismatch between argument types (some are raw pointers, and others are shared_ptr). If I understand your comment, what you have asked for requires a raw -> shared_ptr conversion, which is not possible.

Sorry, I didn't notice that wehn commenting, let me check your changes.

Reply:
@asymingt sorry that my comment was confusing.
I simplified the changes in https://github.com/ros2/rclcpp/pull/1975/commits/5c129b8b17076cc9af55fc9b278ad283c6abb9b3 and https://github.com/ros2/rclcpp/pull/1975/commits/ff241d40333d31912f31570e915d0cb7665ea376